components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java

```diff
@@ -300,11 +311,77 @@ private void startOrganizationBoundTenantFlow(String authorizedOrganization) {
                 .setUserResidentOrganizationId(userResidentOrganizationId);
         PrivilegedCarbonContext.getThreadLocalCarbonContext().setOrganizationId(authorizedOrganization);
         try {
-            String authorizedTenantDomain = AuthorizationValveServiceHolder.getInstance().getOrganizationManager()
-                    .resolveTenantDomain(authorizedOrganization);
+            String authorizedTenantDomain = getOrganizationManager().resolveTenantDomain(authorizedOrganization);
             PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(authorizedTenantDomain, true);
+            populateOrganizationInIdentityContext(authorizedOrganization);
         } catch (OrganizationManagementException e) {
             throw new AuthRuntimeException("Error while resolving tenant domain by organization.", e);
         }
     }
+
+    private void populateRootOrganizationInIdentityContext() {
+
+        int tenantId = IdentityContext.getThreadLocalIdentityContext().getTenantId();
+        if (MultitenantConstants.SUPER_TENANT_ID == tenantId) {
+            IdentityContext.getThreadLocalIdentityContext().setRootOrganization(new RootOrganization.Builder()
+                    .id(MultitenantConstants.SUPER_TENANT_ID)
+                    .name(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)
+                    .organizationId(OrganizationManagementConstants.SUPER_ORG_ID)
+                    .build());
+            return;
+        }
+
+        try {
+            Tenant tenant = IdentityCoreServiceDataHolder.getInstance().getRealmService().getTenantManager()
+                    .getTenant(tenantId);
+            if (tenant == null) {
+                log.debug("Tenant with id: " + tenantId + " does not exist. Cannot initialize root organization.");
+                return;
+            }
+
+            IdentityContext.getThreadLocalIdentityContext().setRootOrganization(new RootOrganization.Builder()
+                    .id(tenantId)
+                    .name(IdentityContext.getThreadLocalIdentityContext().getTenantDomain())
+                    .organizationId(tenant.getAssociatedOrganizationUUID())
+                    .build());
+        } catch (UserStoreException e) {
+            log.error("Error while retrieving tenant with id: " + tenantId, e);
+        }
+    }
```

Comment:
Is this right if you consider a sub org flow

Reply:
Resolving root organization logic is removed from this valve level. 
Fixed with https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/324/commits/922f889076cb9d544d9d67da92488419a15a7f05